x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/bulk_actions/use_custom_rules_count.ts

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useState, useCallback } from 'react';
+import { fetchRules } from '../../../../../containers/detection_engine/rules/api';
+import type { FilterOptions } from '../../../../../containers/detection_engine/rules/types';
+
+interface UseCustomRulesCount {
+  fetchCustomRulesCount: (filterOptions: FilterOptions) => Promise<{ customRulesCount: number }>;
+  customRulesCount: number;
+  isCustomRulesCountLoading: boolean;
+}
+
+export const useCustomRulesCount = (): UseCustomRulesCount => {
+  const [customRulesCount, setCustomRulesCount] = useState<number>(0);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  const fetchCustomRulesCount = useCallback(async (filterOptions: FilterOptions) => {
+    const abortController = new AbortController();
```

Comment:
Let's use `react-query` instead; it helps greatly reduce boilerplate and forget about managing stuff like `AbortController` or `isLoading` states.

Reply:
Doesn't look as best solution in this case:

I need to query data on demand with different parameters every time (filterOptions) and get results from fetch method promise `fetchCustomRulesCount`.
From what I see so far in react-query documentation there is no easy way to do it.

1. It fetches data by on mounting by default. As I don't need it, I have to pass `enabled:false`
2. Then If I need to get data I have to use exposed method `refetch`.
3. Problem is: refetch doesn't support parameters.(https://github.com/tannerlinsley/react-query/discussions/442 - described workarounds)
4. So I would need to introduce state in this hook, to save these parameters there(`filterOptions`) and then pass them into fetch request function(`findRules`) within useQuery. 
6.  But, even so I do this: refetch will return result for parameters from previous state not for parameters it was called before

Other way: is to use useMutation for this purpose(I've seen some recommendations on that), but it doesn't look right to me.

For reference, implementation described above

```typescript
export const useCustomRulesCount = (): UseCustomRulesCount => {
  const [filterOptions, setFilterOptions] = useState<FilterOptions>();
  const { data, isRefetching, refetch } = useQuery<{ customRulesCount: number }>(
    ['customRulesCount'],
    async ({ signal }) => {
      if (!filterOptions) {
        return {
          customRulesCount: 0,
        };
      }
      const res = await fetchRules({
        pagination: { perPage: 1, page: 1 },
        filterOptions: { ...filterOptions, showCustomRules: true },
        signal,
      });

      return {
        customRulesCount: res.total,
      };
    },
    {
      enabled: false,
    }
  );

  const fetchCustomRulesCount = useCallback(
    async (newFilterOptions: FilterOptions) => {
      setFilterOptions(newFilterOptions);

      const { data: newData } = await refetch();

      return {
        customRulesCount: newData?.customRulesCount ?? 0,
      };
    },
    [refetch]
  );

  return {
    fetchCustomRulesCount,
    customRulesCount: data?.customRulesCount ?? 0,
    isCustomRulesCountLoading: isRefetching,
  };
};

```
So, I would rather leave current implementation in place. Happy to discuss any other approaches if we still want to keep useQuery.

Reply:
Hmm, why do you need an imperative handler (`fetchCustomRulesCount`)? Can't we return the number of custom rules from the hook?